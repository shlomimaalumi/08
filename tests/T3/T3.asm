//filename:  Main
@256
D=A
@SP
M=D
//************************write call*****************
//save last values before call to function
@return.Sys.init$48969
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Sys.init
0;JMP
(return.Sys.init$48969)
//******write function Main.T3 with 0 locals ******
//set 0 n_vars  forMain.T3
(Main.T3)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Main.T3199225360)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals199225360
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Main.T3199225360
0;JMP
//*************finish to make n_vars:************
(end.locals199225360)
//************************write call*****************
//save last values before call to function
@return.Main.t1$42339
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Main.t1
0;JMP
(return.Main.t1$42339)
//************************write call*****************
//save last values before call to function
@return.Main.t2$40917
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Main.t2
0;JMP
(return.Main.t2$40917)
//************************write call*****************
//save last values before call to function
@return.Main.t3$97290
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Main.t3
0;JMP
(return.Main.t3$97290)
//************************write call*****************
//save last values before call to function
@return.Main.t4$66103
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Main.t4
0;JMP
(return.Main.t4$66103)
//******write label Main.T3$WHILE******
(Main.Main.T3Main.T3$WHILE)
//******write goto Main.T3$WHILE******
@Main.Main.T3Main.T3$WHILE
0;JMP
//************************write return*****************
@LCL
D=M
@frame
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@ret.address
M=D
@SP
A=M-1
D=M
@ARG
A=M
M=D
@ARG
D=M
@SP
M=D
M=M+1
@frame
A=M
A=A-1
D=M
@THAT
M=D
@frame
A=M
A=A-1
A=A-1
D=M
@THIS
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
D=M
@ARG
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@LCL
M=D
//*************END FUN! we will BACK to:************
@ret.address
A=M
0;JMP
//******write function Main.t1 with 0 locals ******
//set 0 n_vars  forMain.t1
(Main.t1)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Main.t1199225361)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals199225361
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Main.t1199225361
0;JMP
//*************finish to make n_vars:************
(end.locals199225361)
// push constant
@32767
D=A
@SP
A=M
M=D

@SP
M=M+1
// neg
@SP
A=M-1
D=M
M=0
M=M-D
// push constant
@1
D=A
@SP
A=M
M=D

@SP
M=M+1
// sub
@SP
A=M-1
A=A-1
D=M
A=A+1
D=D-M
A=A-1
M=D
@SP
M=M-1
// push constant
@32767
D=A
@SP
A=M
M=D

@SP
M=M+1
// lt
@SP
A=M-1
D=M
@yp0
D;JGE
@SP
A=M-1
A=A-1
D=M
@xpym0
D;JGE
@xys0
0;JMP
(yp0)
@SP
A=M-1
A=A-1
D=M
@xys0
D;JGE
@SP
M=M-1
M=M-1
A=M
M=0
M=M-1
@end0
0;JMP
(xpym0)
@SP
M=M-1
M=M-1
A=M
M=0
@end0
0;JMP
(xys0)
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
D=D-M
M=0
@end0
D;JLE
@SP
A=M
M=M-1
(end0)
@SP
M=M+1

//************************write return*****************
@LCL
D=M
@frame
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@ret.address
M=D
@SP
A=M-1
D=M
@ARG
A=M
M=D
@ARG
D=M
@SP
M=D
M=M+1
@frame
A=M
A=A-1
D=M
@THAT
M=D
@frame
A=M
A=A-1
A=A-1
D=M
@THIS
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
D=M
@ARG
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@LCL
M=D
//*************END FUN! we will BACK to:************
@ret.address
A=M
0;JMP
//******write function Main.t2 with 0 locals ******
//set 0 n_vars  forMain.t2
(Main.t2)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Main.t2199225362)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals199225362
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Main.t2199225362
0;JMP
//*************finish to make n_vars:************
(end.locals199225362)
// push constant
@32767
D=A
@SP
A=M
M=D

@SP
M=M+1
// push constant
@32767
D=A
@SP
A=M
M=D

@SP
M=M+1
// neg
@SP
A=M-1
D=M
M=0
M=M-D
// push constant
@1
D=A
@SP
A=M
M=D

@SP
M=M+1
// sub
@SP
A=M-1
A=A-1
D=M
A=A+1
D=D-M
A=A-1
M=D
@SP
M=M-1
// gt
@SP
A=M-1
D=M
@yp0
D;JGE
@SP
A=M-1
A=A-1
D=M
@xpym0
D;JGE
@xys0
0;JMP
(yp0)
@SP
A=M-1
A=A-1
D=M
@xys0
D;JGE
@SP
M=M-1
M=M-1
A=M
M=0
@end0
0;JMP
(xpym0)
@SP
M=M-1
M=M-1
A=M
M=0
M=M-1
@end0
0;JMP
(xys0)
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
D=D-M
M=0
@end0
D;JGE
@SP
A=M
M=M-1
(end0)
@SP
M=M+1

//************************write return*****************
@LCL
D=M
@frame
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@ret.address
M=D
@SP
A=M-1
D=M
@ARG
A=M
M=D
@ARG
D=M
@SP
M=D
M=M+1
@frame
A=M
A=A-1
D=M
@THAT
M=D
@frame
A=M
A=A-1
A=A-1
D=M
@THIS
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
D=M
@ARG
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@LCL
M=D
//*************END FUN! we will BACK to:************
@ret.address
A=M
0;JMP
//******write function Main.t3 with 0 locals ******
//set 0 n_vars  forMain.t3
(Main.t3)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Main.t3199225363)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals199225363
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Main.t3199225363
0;JMP
//*************finish to make n_vars:************
(end.locals199225363)
// push constant
@20000
D=A
@SP
A=M
M=D

@SP
M=M+1
// neg
@SP
A=M-1
D=M
M=0
M=M-D
// push constant
@1
D=A
@SP
A=M
M=D

@SP
M=M+1
// sub
@SP
A=M-1
A=A-1
D=M
A=A+1
D=D-M
A=A-1
M=D
@SP
M=M-1
// push constant
@30000
D=A
@SP
A=M
M=D

@SP
M=M+1
// gt
@SP
A=M-1
D=M
@yp1
D;JGE
@SP
A=M-1
A=A-1
D=M
@xpym1
D;JGE
@xys1
0;JMP
(yp1)
@SP
A=M-1
A=A-1
D=M
@xys1
D;JGE
@SP
M=M-1
M=M-1
A=M
M=0
@end1
0;JMP
(xpym1)
@SP
M=M-1
M=M-1
A=M
M=0
M=M-1
@end1
0;JMP
(xys1)
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
D=D-M
M=0
@end1
D;JGE
@SP
A=M
M=M-1
(end1)
@SP
M=M+1

//************************write return*****************
@LCL
D=M
@frame
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@ret.address
M=D
@SP
A=M-1
D=M
@ARG
A=M
M=D
@ARG
D=M
@SP
M=D
M=M+1
@frame
A=M
A=A-1
D=M
@THAT
M=D
@frame
A=M
A=A-1
A=A-1
D=M
@THIS
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
D=M
@ARG
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@LCL
M=D
//*************END FUN! we will BACK to:************
@ret.address
A=M
0;JMP
//******write function Main.t4 with 0 locals ******
//set 0 n_vars  forMain.t4
(Main.t4)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Main.t4199225364)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals199225364
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Main.t4199225364
0;JMP
//*************finish to make n_vars:************
(end.locals199225364)
// push constant
@20000
D=A
@SP
A=M
M=D

@SP
M=M+1
// push constant
@30000
D=A
@SP
A=M
M=D

@SP
M=M+1
// neg
@SP
A=M-1
D=M
M=0
M=M-D
// push constant
@1
D=A
@SP
A=M
M=D

@SP
M=M+1
// sub
@SP
A=M-1
A=A-1
D=M
A=A+1
D=D-M
A=A-1
M=D
@SP
M=M-1
// gt
@SP
A=M-1
D=M
@yp2
D;JGE
@SP
A=M-1
A=A-1
D=M
@xpym2
D;JGE
@xys2
0;JMP
(yp2)
@SP
A=M-1
A=A-1
D=M
@xys2
D;JGE
@SP
M=M-1
M=M-1
A=M
M=0
@end2
0;JMP
(xpym2)
@SP
M=M-1
M=M-1
A=M
M=0
M=M-1
@end2
0;JMP
(xys2)
@SP
M=M-1
A=M
D=M
@SP
M=M-1
A=M
D=D-M
M=0
@end2
D;JGE
@SP
A=M
M=M-1
(end2)
@SP
M=M+1

//************************write return*****************
@LCL
D=M
@frame
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@ret.address
M=D
@SP
A=M-1
D=M
@ARG
A=M
M=D
@ARG
D=M
@SP
M=D
M=M+1
@frame
A=M
A=A-1
D=M
@THAT
M=D
@frame
A=M
A=A-1
A=A-1
D=M
@THIS
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
D=M
@ARG
M=D
@frame
A=M
A=A-1
A=A-1
A=A-1
A=A-1
D=M
@LCL
M=D
//*************END FUN! we will BACK to:************
@ret.address
A=M
0;JMP
(endendend123123)
@endendend123123
0;JMP
//filename:  Sys
//******write function Sys.init with 0 locals ******
//set 0 n_vars  forSys.init
(Sys.init)
//iteration number: 0
@0
D=A
@vars
M=D
@ind.for.loop
M=0
(loop.Sys.init167635343)
@ind.for.loop
D=M
@vars
D=D-M
@end.locals167635343
D;JEQ
@SP
A=M
M=0
@ind.for.loop
M=M+1
@SP
M=M+1
@loop.Sys.init167635343
0;JMP
//*************finish to make n_vars:************
(end.locals167635343)
//************************write call*****************
//save last values before call to function
@return.Main.T3$11002
D=A
@SP
A=M
M=D
@SP
M=M+1
@LCL
D=M
@SP
A=M
M=D
@SP
M=M+1
@ARG
D=M
@SP
A=M
M=D
@SP
M=M+1
@THIS
D=M
@SP
A=M
M=D
@SP
M=M+1
@THAT
D=M
@SP
A=M
M=D
@SP
M=M+1
@SP
D=M
D=D-1
D=D-1
D=D-1
D=D-1
D=D-1
@ARG
M=D
@SP
D=M
@LCL
M=D

//*************START FUN! we will jump to:************
@Main.T3
0;JMP
(return.Main.T3$11002)
//******write label Sys.init$WHILE******
(Sys.Sys.initSys.init$WHILE)
//******write goto Sys.init$WHILE******
@Sys.Sys.initSys.init$WHILE
0;JMP
(endendend123123)
@endendend123123
0;JMP
