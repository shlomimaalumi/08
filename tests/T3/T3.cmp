| RAM[266] | RAM[267] | RAM[268] | RAM[269] |
|      -1  |      -1  |       0  |      -1  |
