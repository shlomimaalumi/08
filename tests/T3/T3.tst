// This file is part of the materials accompanying the book
// "The Elements of Computing Systems" by Nisan and Schocken, 
// MIT Press 2004. Book site: http://www.idc.ac.il/tecs
// File name: projects/07/StackArithmetic/StackTest/StackTest.tst.
// Version: beta 1.4.

load T3.asm,
output-file T3.out,
compare-to T3.cmp,
output-list RAM[266]%D2.6.2 RAM[267]%D2.6.2 
            RAM[268]%D2.6.2 RAM[269]%D2.6.2;

repeat 10000 {
  ticktock;
}

output;
