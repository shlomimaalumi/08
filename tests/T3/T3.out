| RAM[266] | RAM[267] | RAM[268] | RAM[269] |
